docs/dg/dev/guidelines/performance-guidelines/search-performance-guidelines.md

```diff
@@ -0,0 +1,488 @@
+
+```json
+{
+  "query": {
+    "bool": {
+      "filter": [
+        {"term": {"category": "electronics"}}
+      ]
+    }
+  },
+  "aggregations": {
+    "filtered_brands": {
+      "terms": {
+        "field": "brand"
+      }
+    }
+  }
+}
+```
+
+### When global aggregations might be acceptable
+
+Global aggregations are only acceptable in specific scenarios:
+
+- **Administrative dashboards**: Where performance is less critical and global statistics are genuinely needed.
+- **Background jobs**: For generating reports or analytics outside of user-facing requests.
+- **Small datasets**: On indices with a limited number of documents (typically under 10,000).
+
+In all cases, consider caching the results and updating them periodically rather than calculating on every request.
+
+## Avoid deep pagination: search is search, not a catalog
+
+Elasticsearch and OpenSearch have a default maximum of 10,000 documents that can be retrieved using the `from` and `size` parameters (offset-based pagination). This limit exists for good performance reasons.
+
+### Why deep pagination is problematic
+
+- **Resource consumption**: To retrieve page 500 (items 10,000-10,020), Elasticsearch must find and sort 10,020 documents, then discard the first 10,000. This is extremely inefficient.
+- **Memory pressure**: Deep pagination requires holding large result sets in memory across shards.
+- **Distributed search overhead**: In a multi-shard environment, each shard must return `from + size` results to the coordinating node, which then sorts and returns only `size` results.
+
+### What to avoid
+
+Do not allow users to paginate deeply through search results:
+
+```json
+{
+  "from": 9990,
+  "size": 20,
+  "query": {...}
+}
+```
+
+Attempting to exceed the 10,000 document limit results in errors:
+
+```text
+Result window is too large, from + size must be less than or equal to: [10000]
+```
+
+While you can increase `index.max_result_window`, this is strongly discouraged as it degrades performance.
+
+### Search is not a catalog browser
+
+Important principle: **Search should help users find specific products, not browse the entire catalog**.
+
+Users who reach page 50+ of search results are not finding what they need. Instead of enabling deep pagination:
+
+1. **Improve search relevance**: Better search algorithms and ranking help users find products on the first few pages.
+2. **Suggest query refinement**: Guide users to apply filters and narrow their search.
+3. **Use faceted navigation**: Let users filter by categories, attributes, and price ranges.
+4. **Implement sort options**: Allow sorting by relevance, price, rating, or newness.
+
+
+
+### Alternative approaches for deep access
+
+If you genuinely need to access results beyond 10,000 documents:
+
+**1. Search After (recommended for sequential access)**
+
+Use the `search_after` parameter for efficient pagination through large result sets:
```

Comment:
Do you know if it's easy to implement it on Spryker?
I know that some customers failed to do this